File: docs-web.resources/com.ibm.docs.web.resources/WebContent/js/wseditor/js/model/RulesObject/RulesObj.js
/* ***************************************************************** */
/*                                                                   */
/* IBM Confidential                                                  */
/*                                                                   */
/* IBM Docs Source Materials                                         */
/*                                                                   */
/* (c) Copyright IBM Corporation 2012. All Rights Reserved.          */
/*                                                                   */
/* U.S. Government Users Restricted Rights: Use, duplication or      */
/* disclosure restricted by GSA ADP Schedule Contract with IBM Corp. */
/*                                                                   */
/* ***************************************************************** */

dojo.provide("websheet.model.RulesObject.RulesObj");
dojo.declare("websheet.model.RulesObject.RulesObj", null, {
	_doc: null,
	_id: null,			//string, same as rangeId defined in range
	_ranges: null, 		//data validation applied range
	
	_topRow: null,
	_leftCol: null,
	
	VALUETYPE: websheet.Constant.valueType,
	
	constructor:function(range, doc, criteria){
		this._doc = doc;
		this._ranges = [];
		this._ranges.push(range);
		this._id = range._id;
	},
	
	createNewInstance : function(range){
		var rulesObj = this.constructor.apply(null,[range, this._doc]);
		
		rulesObj._topRow = range.getStartRow();
		rulesObj._leftCol = range.getStartCol();
		
		return rulesObj;
	},
	
	rangeNum: function()
	{
		return this._ranges.length;
	},
	
	getRanges: function()
	{
		return this._ranges;
	},
	
	addRange:function(range, bParse){
		this._ranges.push(range);
		if(bParse)
			this.parse4Range(range);
	},
	
	removeRange: function(range){
		var len = this._ranges.length;
		var i = 0;
		for(; i < len; i++){
			if(this._ranges[i].getId() == range.getId()){
				this.clearAll(range.getId());
				this._ranges.splice(i,1);
				break;
			}
		}
		if(this._ranges.length == 0){
			this._doc.removeRulesObjByUsage(this._id, this._usage);
			this._doc._areaManager.deleteAreaInUsageMap(range);
			return;
		}
		if(i == 0){//set the range id of the second one to parentid;
			this._doc.renameRulesObjByUsage(this, this._ranges[0]._id, this._usage);
		}
	},

	getJSON4Range: function(rangeInfo, baseRange)
	{
		
	},
	
	//For copy paste, row repeat or col repeat
	parse4Range: function(range){
		this._parse(range);
	},
	
	_parse: function(sharedFormulaRef, bFirst, tokenList, areaManager){
		if(!tokenList){
			tokenList = this._getTokenList();
			if(tokenList.length == 0)
				return;
			this._updateTokenArrays(tokenList);
		}
		
		if(!areaManager)
			areaManager = this._doc.getAreaManager();
		
		var deltaR = sharedFormulaRef.getStartRow() - this._topRow;
		var deltaC = sharedFormulaRef.getStartCol() - this._leftCol;
		var rowNum = sharedFormulaRef.getEndRow() - sharedFormulaRef.getStartRow();
		var colNum = sharedFormulaRef.getEndCol() - sharedFormulaRef.getStartCol();
		for(var j = 0; j < tokenList.length; j++){
			var token = tokenList[j];
			var ref = token.getValue();
			if(token.getTokenType() == websheet.parse.tokenType.RANGEREF_TOKEN && websheet.Helper.isRelativeRef(token.getRefMask()) && ref.isValid()){
				if(!ref.sheetName){
					var sheetName = sharedFormulaRef.getSheetName();
					ref.setSheetName(sheetName);
				}
				
				var startRow = ref.startRow;
				var endRow = ref.endRow;
				var startCol = ref.startCol;
				var endCol = ref.endCol;
				
				var rowSize = endRow - startRow + 1;
				var colSize = endCol - startCol + 1;
				
				if((ref.refMask & websheet.Constant.RefAddressType.ROW) > 0 && (ref.refMask & websheet.Constant.RefAddressType.ABS_ROW) == 0)
					startRow += deltaR;
				if((ref.refMask & websheet.Constant.RefAddressType.COLUMN) > 0 && (ref.refMask & websheet.Constant.RefAddressType.ABS_COLUMN) == 0)
					startCol += deltaC;
				var isCell = (ref.refMask & websheet.Constant.RefAddressType.END_ROW) == 0 && (ref.refMask & websheet.Constant.RefAddressType.END_COLUMN) == 0;	
				if(isCell){
					endRow = startRow;
					endCol = startCol;
				}else{
					if((ref.refMask & websheet.Constant.RefAddressType.END_ROW) > 0 && (ref.refMask & websheet.Constant.RefAddressType.ABS_END_ROW) == 0)
						endRow += deltaR;
					if((ref.refMask & websheet.Constant.RefAddressType.END_COLUMN) > 0 && (ref.refMask & websheet.Constant.RefAddressType.ABS_END_COLUMN) == 0)
						endCol += deltaC;
				}
				
				//var mask = websheet.Constant.RANGE_MASK;
				if(((ref.refMask & websheet.Constant.RefAddressType.ROW) > 0 && (ref.refMask & websheet.Constant.RefAddressType.ABS_ROW) == 0)
						||((ref.refMask & websheet.Constant.RefAddressType.END_ROW) > 0 && (ref.refMask & websheet.Constant.RefAddressType.ABS_END_ROW) == 0)){
					if(endRow < startRow){
						var tmp = startRow;
						startRow = endRow;
						endRow = tmp;
					}
					endRow += rowNum;
				}
				if(((ref.refMask & websheet.Constant.RefAddressType.COLUMN) > 0 && (ref.refMask & websheet.Constant.RefAddressType.ABS_COLUMN) == 0)
						||((ref.refMask & websheet.Constant.RefAddressType.END_COLUMN) > 0 && (ref.refMask & websheet.Constant.RefAddressType.ABS_END_COLUMN) == 0)){
					if(endCol < startCol){
						var tmp = startCol;
						startCol = endCol;
						endCol = tmp;
					}
					endCol += colNum;
				}
				
				var mask = ref.refMask;
				if(isCell){//cell, maybe need to expand to range.
					if(startRow != endRow || startCol != endCol){
						if((ref.refMask & websheet.Constant.RefAddressType.ABS_ROW) > 0)
							mask |= websheet.Constant.RefAddressType.ABS_END_ROW;
						mask |=  websheet.Constant.RefAddressType.END_ROW;
						if((ref.refMask & websheet.Constant.RefAddressType.ABS_COLUMN) > 0)
							mask |= websheet.Constant.RefAddressType.ABS_END_COLUMN;
						mask |= websheet.Constant.RefAddressType.END_COLUMN;
					}
				}
				
				var sharedReferenceRef = new websheet.parse.SharedReferenceRef(new websheet.parse.ParsedRef(ref.sheetName, startRow, startCol, endRow, endCol, mask), rowSize, colSize);
				var newSharedRefToken = new websheet.parse.RefToken();
				dojo.mixin(newSharedRefToken, token);
	            newSharedRefToken.setValue(sharedReferenceRef);
	            newSharedRefToken.setRefMask(mask);
	            sharedFormulaRef.pushRefToken(newSharedRefToken, true);
	            areaManager.startListeningArea(sharedReferenceRef._parsedRef, sharedFormulaRef, sharedReferenceRef);
			}else{
				var refToken = bFirst ? this._generateRefToken(token) : token;
				var area = refToken.getValue();
				if(!area.isInstanceOf(websheet.parse.Area)){
					refToken = this._generateRefToken(token);
					area = refToken.getValue();
				}
				sharedFormulaRef.pushRefToken(refToken, true);
				area.addListener(sharedFormulaRef);
			}
		}
	},
	
	//generate ref tokens for sharedformularef.
	parse:function(){
		var tokenList = this._getTokenList(true);		
		if(tokenList.length == 0)
			return;
		
		this._parseBaseRef();
		var areaManager = this._doc.getAreaManager();
		
		for(var i = 0; i < this._ranges.length; i++){
			var sharedFormulaRef = this._ranges[i];
			this._parse(sharedFormulaRef, i == 0, tokenList, areaManager);
		}
	},
	
	_getTokenList: function()
	{
		
	},
	
	_updateTokenArrays: function(tokenList){
		var range = this._ranges[0];
		var deltaR = range.getStartRow() - this._topRow;
		var deltaC = range.getStartCol() - this._leftCol;
		var rowNum = range.getEndRow() - range.getStartRow();
		var colNum = range.getEndCol() - range.getStartCol();
		
		var refTokens = range.getRefTokens();
		for(var i = 0; i < refTokens.length; i ++){
			var refToken = refTokens[i];
			var ref = refToken.getValue();
			if(ref.usage != websheet.Constant.RangeUsage.SHARED_REFS)
				continue;
			var token = tokenList[i];
			var parsedRef = token.getValue();
			
			//parsedRef
			var startRow = ref.getStartRow();
			var endRow = ref.getEndRow();
			var startCol = ref.getStartCol();
			var endCol = ref.getEndCol();
			
			if((parsedRef.refMask & websheet.Constant.RefAddressType.ROW) > 0 && (parsedRef.refMask & websheet.Constant.RefAddressType.ABS_ROW) == 0)
				startRow -= deltaR;
			if((parsedRef.refMask & websheet.Constant.RefAddressType.COLUMN) > 0 && (parsedRef.refMask & websheet.Constant.RefAddressType.ABS_COLUMN) == 0)
				startCol -= deltaC;			
			if(parsedRef.refMask == websheet.Constant.DEFAULT_CELL_MASK || parsedRef.refMask == websheet.Constant.CELL_MASK){
				endRow = startRow;
				endCol = startCol;
			}else{
				if((parsedRef.refMask & websheet.Constant.RefAddressType.END_ROW) > 0 && (parsedRef.refMask & websheet.Constant.RefAddressType.ABS_END_ROW) == 0)
					endRow -= deltaR;
				if((parsedRef.refMask & websheet.Constant.RefAddressType.END_COLUMN) > 0 && (parsedRef.refMask & websheet.Constant.RefAddressType.ABS_END_COLUMN) == 0)
					endCol -= deltaC;
			}
			
			parsedRef.startRow = startRow;
			parsedRef.endRow = endRow;
			parsedRef.startCol = startCol;
			parsedRef.endCol = endCol;
		}
	},
	
	_generateRefToken: function(token){
		var areaManager = this._doc.getAreaManager();
		
		var range = this._ranges[0];
		var sheetName = range.getSheetName();
		if(token.getTokenType() == websheet.parse.tokenType.NAME){
			var name = token.getName();
			var area = areaManager.startListeningNameArea(name, null);
			 if(area){
				token.setValue(area);
				token.setName(area.getId());
				if (area.getUsage() == websheet.Constant.RangeUsage.NAME) {
					if(!area.isValid())
						token._error = websheet.Constant.ERRORCODE["524"];
					else if(area.setNeedPartial && area.getSheetName() != sheetName)
						area.setNeedPartial(true);
				} else {
					 token.setValue(area);
	           		 token._error = websheet.Constant.ERRORCODE["525"];//#NAME!
				}
			 }
		}else{
			var parsedRef = token.getValue();
			if(!parsedRef.sheetName)
				parsedRef.setSheetName(sheetName);
			var area = areaManager.startListeningArea(parsedRef, null);
			if(area.getSheetName() != sheetName)
				area.setNeedPartial(true);
			token.setValue(area);
			token.setRefMask(parsedRef.refMask);
		}
		return token;
	},
	
	_parseBaseRef: function(){
		var topRow, leftCol;
		for(var i = 0; i < this._ranges.length; i++){
			var range = this._ranges[i];
			if(!topRow || topRow > range.getStartRow())
				topRow = range.getStartRow();
			if(!leftCol || leftCol > range.getStartCol())
				leftCol = range.getStartCol();
		}
		this._topRow = topRow;
		this._leftCol = leftCol;
	},
	
	getTokenIdxs: function(ref, tokenArray, tokenIdxs)
	{
		
	},
	
	isSplitable: function()
	{
		return true;		
	},
	
	//clear all cached data for the value of the range
	clearAllData: function(rangeId, dirtyFlag)
	{
		
	},
	
	//clear partial cached data for the value of the range
	clearData4Cell: function(rangeId, sr, sc, er, ec, dirtyFlag)
	{
		
	},
	
	//clear all cached data for range
	clearAll: function(rangeId)
	{
		
	}
});

File: docsapp/com.ibm.concord.spreadsheet/src/com/ibm/concord/spreadsheet/document/model/rulesObject/RulesObj.java
package com.ibm.concord.spreadsheet.document.model.rulesObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import java.util.logging.Logger;

import com.ibm.concord.spreadsheet.document.model.formula.Area;
import com.ibm.concord.spreadsheet.document.model.formula.AreaManager;
import com.ibm.concord.spreadsheet.document.model.formula.FormulaUtil;
import com.ibm.concord.spreadsheet.document.model.formula.Reference;
import com.ibm.concord.spreadsheet.document.model.formula.SharedFormulaRef4RulesObj;
import com.ibm.concord.spreadsheet.document.model.formula.SharedReferenceRef;
import com.ibm.concord.spreadsheet.document.model.formula.FormulaUtil.FormulaToken;
import com.ibm.concord.spreadsheet.document.model.formula.FormulaUtil.ReferenceToken;
import com.ibm.concord.spreadsheet.document.model.impl.Document;
import com.ibm.concord.spreadsheet.document.model.impl.Sheet;
import com.ibm.concord.spreadsheet.document.model.impl.Range.RangeInfo;
import com.ibm.concord.spreadsheet.document.model.util.ModelHelper;
import com.ibm.json.java.JSONObject;
import com.ibm.concord.spreadsheet.common.ConversionConstant;
import com.ibm.concord.spreadsheet.common.ReferenceParser;
import com.ibm.concord.spreadsheet.common.ReferenceParser.ParsedRefType;
import com.ibm.concord.spreadsheet.conversion.impl.ConversionUtil.RangeUsage;

public abstract class RulesObj
{
	private static final Logger LOG = Logger.getLogger(RulesObj.class.getName());
	protected Document doc;
	protected String id;
	protected List<SharedFormulaRef4RulesObj> ranges = null;
	protected RangeUsage usage;
	protected int topRow;
	protected int leftCol;
	private Map<String, ReferenceParser.ParsedRef> parsedRefCache;
	
	public RulesObj(SharedFormulaRef4RulesObj area, Document doc){
		this.doc = doc;
		
		ranges = new ArrayList<SharedFormulaRef4RulesObj>();
		ranges.add(area);
		
		id = area.getId();
		
	}
	
	public RangeUsage getUsage()
	{
		return usage;
	}
	
	public List<SharedFormulaRef4RulesObj> ranges()
	{
		return ranges;
	}
	
	public void addRange(SharedFormulaRef4RulesObj range)
	{
		ranges.add(range);
	}
	
	public int rangeNum()
	{
		return ranges.size();
	}
	
	public void removeRange(SharedFormulaRef4RulesObj range)
	{
		int len = ranges.size();
		int i = 0;
		for(i = 0; i < len; i++){
			if(ranges.get(i).getId() == range.getId()){
				ranges.remove(i);
				break;
			}
		}
		if(ranges.size() == 0){
			doc.removeRulesObj(id);
			return;
		}
		if(i == 0){
			String newName = ranges.get(0).getId();
			doc.renameRulesObj(this, newName, usage);
			id = newName;
		}
	}
	
	 public abstract RulesObj createNewInstance(SharedFormulaRef4RulesObj range, ArrayList<FormulaToken> refTokens);
	
	 public String getId()
	 {
		 return id;
	 }
	 
	 protected void parseBaseRef()
	 {
		topRow = 0;
		leftCol = 0;
		for(int i = 0; i < this.ranges.size(); i++){
			SharedFormulaRef4RulesObj range = ranges.get(i);
			if(topRow == 0 || topRow > range.getStartRow())
				topRow = range.getStartRow();
			if(leftCol == 0|| leftCol > range.getStartCol())
				leftCol = range.getStartCol();
		}
	 }
	 
	 protected List<FormulaToken> getTokenList(boolean setTokenPos)
	 {
		 return null;
	 }
	 
	 protected List<FormulaToken> getTokenList()
	 {
		 return null;
	 }
	 
	//Only add sharedreftokens to sharedformularef, abslate refs and names only in datavalidaion's tokenarray.
	 public void parse()
	 {	 
		 List<FormulaToken> tokenList = getTokenList(true);
		 if(tokenList.size() == 0)
			 return;
		 
		 parseBaseRef();
		 AreaManager areaMgr = doc.getAreaManager();
		 
		 for(int i = 0; i < ranges.size(); i++)
		 {
			 SharedFormulaRef4RulesObj range = ranges.get(i);
			 parse(range, i == 0, tokenList, areaMgr);
		 }
	 }
	 
	 public void parse4Range(SharedFormulaRef4RulesObj area)
	{
		parse(area, false, null, null);
	}
	 
	 private void updateTokenArrays(List<FormulaToken> tokenList)
	 {
		SharedFormulaRef4RulesObj range = ranges.get(0);
		int deltaR = range.getStartRow() - topRow;
		int deltaC = range.getStartCol() - leftCol;
		int rowNum = range.getEndRow() - range.getStartRow();
		int colNum = range.getEndCol() - range.getStartCol();
		
		 List<FormulaToken> refTokens = range.getTokenList();
		 for(int i = 0; i < refTokens.size(); i ++){
			 FormulaToken refToken = refTokens.get(i);
			 Area area = refToken.getArea();
			 if(!(area instanceof SharedReferenceRef))
				continue;
			 
			 FormulaToken token = tokenList.get(i);
			 String addr = token.getText();
			 ReferenceParser.ParsedRef parsedRef = parsedRefCache != null ? parsedRefCache.get(addr) : null;
			 if(parsedRef == null)
				 parsedRef = ReferenceParser.parse(addr.trim());
			 
			 int startRow = area.getStartRow();
			 int endRow = area.getEndRow();
			 int startCol = area.getStartCol();
			 int endCol = area.getEndCol();
			 
			 if(ParsedRefType.COLUMN != parsedRef.type && (parsedRef.patternMask & ReferenceParser.ABSOLUTE_ROW) == 0)
		    	  startRow -= deltaR;
			 if(ParsedRefType.ROW != parsedRef.type && (parsedRef.patternMask & ReferenceParser.ABSOLUTE_COLUMN) == 0)
				 startCol -= deltaC;
		     if(ParsedRefType.CELL == parsedRef.type)
		     {
		    	 endRow = startRow;
		    	 endCol = startCol;
		     }else{
		    	 if(ParsedRefType.COLUMN != parsedRef.type && (parsedRef.patternMask & ReferenceParser.ABSOLUTE_END_ROW) == 0)
			    	  endRow -= deltaR;
		    	 if(ParsedRefType.ROW != parsedRef.type && (parsedRef.patternMask & ReferenceParser.ABSOLUTE_END_COLUMN) == 0)
			    	 endCol -= deltaC;
		    	 if(ParsedRefType.COLUMN != parsedRef.type && ((parsedRef.patternMask & ReferenceParser.ABSOLUTE_ROW) == 0 || (parsedRef.patternMask & ReferenceParser.ABSOLUTE_END_ROW) == 0))
		    		 endRow -= rowNum;
		    	 if(ParsedRefType.ROW != parsedRef.type && ((parsedRef.patternMask & ReferenceParser.ABSOLUTE_COLUMN) == 0 || (parsedRef.patternMask & ReferenceParser.ABSOLUTE_END_COLUMN) == 0))
		    		endCol -= colNum;
		     }
		     
		     String newAddr = ModelHelper.getAddress(parsedRef.sheetName, startRow, startCol, 
		    		 parsedRef.sheetName, endRow, endCol, parsedRef.patternMask, false);
		     token.setText(newAddr);
		 }
	}
	 
	 private void parse(SharedFormulaRef4RulesObj range, boolean bFirst, List<FormulaToken> tokenList, AreaManager areaMgr)
	{
		if(tokenList == null)
		{
			tokenList = this.getTokenList();
			if(tokenList.size() == 0)
				 return;
			 updateTokenArrays(tokenList);
		}
		if(areaMgr == null)
			areaMgr = doc.getAreaManager();
		
		if(parsedRefCache == null)
			parsedRefCache = new HashMap<String, ReferenceParser.ParsedRef>();
		
		 int deltaR = range.getStartRow() - topRow;
		 int deltaC = range.getStartCol() - leftCol;
		 int rowNum = range.getEndRow() - range.getStartRow();
		 int colNum = range.getEndCol() - range.getStartCol();
		 for(int j = 0; j < tokenList.size(); j++){
			 FormulaToken token = tokenList.get(j);
			 String addr = token.getText();
			 if(token.getType() == FormulaUtil.TokenType.RANGE){
				 ReferenceParser.ParsedRef parsedRef = parsedRefCache.get(addr);
				 if(parsedRef == null)
				 {
					 parsedRef = ReferenceParser.parse(addr.trim());
					 parsedRefCache.put(addr, parsedRef);
					 if(parsedRef.sheetName == null)
			         {
						 int sheetId = range.getSheetId();
						 Sheet s = doc.getSheetById(sheetId);
						 parsedRef.sheetName = s.getSheetName();
			         }
					 if (ParsedRefType.CELL == parsedRef.type){
						 parsedRef.endRow = parsedRef.startRow;
						 parsedRef.endCol = parsedRef.startCol;
					 }
				 }
				 ReferenceToken refToken = (ReferenceToken)token;
				
				 if(parsedRef.isValid() && RulesObjUtil.isRelativeRef(parsedRef.patternMask))
				 {
					 Sheet s = doc.getSheetByName(parsedRef.getSheetName());
					 if(s != null)
					 {
						 int startRow = ParsedRefType.COLUMN == parsedRef.type ? 1 : parsedRef.getIntStartRow();
						 int endRow = ParsedRefType.COLUMN == parsedRef.type ? ConversionConstant.MAX_ROW_NUM : parsedRef.getIntEndRow();
						 int startCol = ParsedRefType.ROW == parsedRef.type ? 1 : parsedRef.getIntStartCol();
						 int endCol = ParsedRefType.ROW == parsedRef.type ? ConversionConstant.MAX_COL_NUM : parsedRef.getIntEndCol();

						 int rowSize = endRow - startRow + 1;
						 int colSize = endCol - startCol + 1;
						 
						 if(ParsedRefType.COLUMN != parsedRef.type && (parsedRef.patternMask & ReferenceParser.ABSOLUTE_ROW) == 0)
					    	  startRow += deltaR;
						 if(ParsedRefType.ROW != parsedRef.type && (parsedRef.patternMask & ReferenceParser.ABSOLUTE_COLUMN) == 0)
							 startCol += deltaC;
					     if(ParsedRefType.CELL == parsedRef.type)
					     {
					    	 endRow = startRow;
					    	 endCol = startCol;
					     }
					     else
					     {
					    	 if(ParsedRefType.COLUMN != parsedRef.type && (parsedRef.patternMask & ReferenceParser.ABSOLUTE_END_ROW) == 0)
						    	  endRow += deltaR;
					    	 if(ParsedRefType.ROW != parsedRef.type && (parsedRef.patternMask & ReferenceParser.ABSOLUTE_END_COLUMN) == 0)
						    	 endCol += deltaC;
					     }
					    
					     if(ParsedRefType.COLUMN != parsedRef.type && ((parsedRef.patternMask & ReferenceParser.ABSOLUTE_ROW) == 0 || (parsedRef.patternMask & ReferenceParser.ABSOLUTE_END_ROW) == 0)){
					    	 if(endRow < startRow){
								int tmp = startRow;
								startRow = endRow;
								endRow = tmp;
							}
					    	 endRow += rowNum;
					     }
						 
					     if(ParsedRefType.ROW != parsedRef.type && ((parsedRef.patternMask & ReferenceParser.ABSOLUTE_COLUMN) == 0 || (parsedRef.patternMask & ReferenceParser.ABSOLUTE_END_COLUMN) == 0)){
					    	 if(endCol < startCol){
								int tmp = startCol;
								startCol = endCol;
								endCol = tmp;
							}
							endCol += colNum;
					     }
					     
					     int mask = parsedRef.patternMask;
					     if(ParsedRefType.CELL == parsedRef.type && (startRow != endRow || startCol != endCol)){
					    	if((mask & ReferenceParser.ABSOLUTE_ROW) > 0)
					    		mask |= ReferenceParser.ABSOLUTE_END_ROW;
					    	mask |= ReferenceParser.END_ROW;
					    	
					    	if((mask & ReferenceParser.ABSOLUTE_COLUMN) > 0)
					    		mask |= ReferenceParser.ABSOLUTE_END_COLUMN;
					    	mask |= ReferenceParser.END_COLUMN;
					     }
					     
					     int sheetId = s.getId();
					     RangeInfo info = new RangeInfo(sheetId, startRow, startCol, endRow, endCol, parsedRef.type);
					     SharedReferenceRef sharedReferenceRef = new SharedReferenceRef(info, doc, rowSize, colSize, parsedRef.patternMask);
					     ReferenceToken sharedRefToken = refToken.copy();
					     sharedRefToken.setArea(sharedReferenceRef);
					     sharedRefToken.setRefMask(mask);
					     areaMgr.startListeningArea(info, range, sharedReferenceRef);
					     range.pushRefToken(sharedRefToken, true);
				     }else
				    	 range.pushRefToken(token, true);
				 }
				 else
				 {
					 Area area = refToken.getArea();
					 if(bFirst || area == null)
					 {
						 Sheet s = doc.getSheetByName(parsedRef.sheetName);
						 if(s != null)
						 {
				        	 int sheetId = s.getId();
							 RangeInfo info = new RangeInfo(sheetId, parsedRef.getIntStartRow(), parsedRef.getIntStartCol(), parsedRef.getIntEndRow(), parsedRef.getIntEndCol(), parsedRef.type);							 
							 area = new Reference(info, doc);
							 refToken.setArea(area);
							 refToken.setRefMask(parsedRef.patternMask);
							 areaMgr.startListeningArea(info, null, area);
						 }
					 }
				 	range.pushRefToken(refToken, true);
				 	if(area != null)
				 		area.addListener(range);
				 }
			 }
			 else
				 range.pushRefToken(token, true);
		 }
	}
	 
	 public List<SharedFormulaRef4RulesObj> getRanges()
	 {
		 ArrayList<SharedFormulaRef4RulesObj> list = new ArrayList<SharedFormulaRef4RulesObj>();
		 SharedFormulaRef4RulesObj range = ranges.get(0);
		 JSONObject data = toJson();		
		 range.setJsonData(data);
		 
		 list.add(range);
		 
		 int len = ranges.size();
		 if(len > 1)
		 {
			 JSONObject d = new JSONObject();
			 d.put(ConversionConstant.PARENTID, id);
			 for(int i = 1; i < len; i++)
			 {
				 SharedFormulaRef4RulesObj r = ranges.get(i);
				 r.setJsonData(d);
				 list.add(r);
			 }
		 }
		 
		 return list;
	 }
	 
	 public JSONObject toJson()
	 {
		 return null;
	 }
	 
	 public boolean isSplitable(boolean bRow)
	 {
		 return true;
	 }
}